File: scala/src/main/scala/com/laamella/code_state_machine/builder/DslStateMachineBuilder.scala
package com.laamella.code_state_machine.builder

import com.laamella.code_state_machine._
import com.laamella.code_state_machine.action.LogAction

import scala.collection.mutable

/**
  * A pretty "DSL" builder for a state machine.
  *
  * @tparam State State type. Each state should have a single instance of this type.
  *           An enum is a good fit.
  * @tparam Event Event type. Events come into the state machine from the outside
  *           world, and are used to trigger state transitions.
  * @tparam Priority Priority type. Will be used to give priorities to transitions.
  *           Enums and Integers are useful here.
  */
class DslStateMachineBuilder[State, Event, Priority <: Ordered[Priority]](defaultPriority: Priority) {
  private val builder = new StateMachineBuilder[State, Event, Priority]()

  class DefiningState(sourceStates: mutable.Set[State]) {
    def except(states: State*): DefiningState = {
      states.foreach(sourceStates.remove)
      this
    }

    def onExit(action: (() => Unit)*): DefiningState = {
      sourceStates.foreach(builder.addExitActions(_, action))
      this
    }

    def onEntry(action: (() => Unit)*): DefiningState = {
      sourceStates.foreach(builder.addEntryActions(_, action))
      this
    }

    def onExitLog(logText: String): DefiningState = onExit(new LogAction(logText))

    def onEntryLog(logText: String): DefiningState = onEntry(new LogAction(logText))

    def isAnEndState: DefiningState = {
      sourceStates.foreach(builder.addEndState)
      this
    }

    def isAStartState: DefiningState = {
      sourceStates.foreach(builder.addStartState)
      this
    }

    def areEndStates: DefiningState = isAnEndState

    def areStartStates: DefiningState = isAStartState

    def when(conditions: Condition[Event]*): DefiningTransition = new DefiningTransition(sourceStates, conditions)

    def always: DefiningTransition = new DefiningTransition(sourceStates, Seq(new AlwaysCondition[Event]()))

    def never: DefiningTransition = new DefiningTransition(sourceStates, Seq(new NeverCondition[Event]()))

    def after(milliseconds: Long): DefiningTransition = new DefiningTransition(sourceStates, Seq(new AfterCondition[Event](milliseconds)))

    def onEvents(events: Event*): DefiningTransition = {
      val condition = if (events.length == 1) {
        val singleEvent = events(0)
        new SingleEventMatchCondition[Event](singleEvent)
      } else {
        new MultiEventMatchCondition[Event](events: _*)
      }
      new DefiningTransition(sourceStates, Seq(condition))
    }

    // FIXME find a way to rescue these conditions
//      def active(statesThatMustBeActive: State*): Condition[Event] = {
//        new StatesActiveCondition[State, Event, Priority](machine, statesThatMustBeActive: _*)
//      }
    //
//      def inactive(statesThatMustBeInactive: State*): Condition[Event] = {
//        new StatesInactiveCondition[State, Event, Priority](machine, statesThatMustBeInactive: _*)
//      }

  }

  class DefiningTransition(sourceStates: mutable.Set[State], conditions: Seq[Condition[Event]]) {
    private val actions = mutable.MutableList[() => Unit]()
    private var priority = defaultPriority

    def doing(action: () => Unit): DefiningTransition = {
      actions += action
      this
    }

    def log(logText: String): DefiningTransition = doing(new LogAction(logText))


    def goTo(destinationState: State): DefiningState = {
      sourceStates.foreach(sourceState => builder.addTransition(new Transition[State, Event, Priority](sourceState, destinationState, conditions, priority, actions)))
      new DefiningState(sourceStates)
    }

    def withPrio(priority: Priority): DefiningTransition = {
      this.priority = priority
      this
    }
  }

  def build(): StateMachine[State, Event, Priority] = builder.build()

  def state(state: State): DefiningState = states(state)

  def states(states: State*): DefiningState = new DefiningState(mutable.HashSet[State](states: _*))
}

File: java/src/main/java/com/laamella/code_state_machine/builder/DslStateMachineBuilder.java
package com.laamella.code_state_machine.builder;

import java.util.Arrays;
import java.util.HashSet;
import java.util.Objects;
import java.util.Set;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.laamella.code_state_machine.Action;
import com.laamella.code_state_machine.Actions;
import com.laamella.code_state_machine.Condition;
import com.laamella.code_state_machine.Conditions;
import com.laamella.code_state_machine.StateMachine;
import com.laamella.code_state_machine.Transition;
import com.laamella.code_state_machine.action.LogAction;
import com.laamella.code_state_machine.condition.AfterCondition;
import com.laamella.code_state_machine.condition.AlwaysCondition;
import com.laamella.code_state_machine.condition.MultiEventMatchCondition;
import com.laamella.code_state_machine.condition.NeverCondition;
import com.laamella.code_state_machine.condition.SingleEventMatchCondition;
import com.laamella.code_state_machine.condition.StatesActiveCondition;
import com.laamella.code_state_machine.condition.StatesInactiveCondition;

import static java.util.Objects.*;

/**
 * A pretty "DSL" builder for a state machine.
 */
public abstract class DslStateMachineBuilder<T, E, P extends Comparable<P>> implements StateMachineBuilder<T, E, P> {
    private static final Logger log = LoggerFactory.getLogger(DslStateMachineBuilder.class);

    public class DefiningState {
        private final Set<T> sourceStates;
        private final StateMachine<T, E, P>.Internals internals;

        public DefiningState(final Set<T> sourceStates, final StateMachine<T, E, P>.Internals internals) {
            this.sourceStates = sourceStates;
            this.internals = internals;
        }

        @SafeVarargs
        public final DefiningState except(final T... states) {
            for (final T state : states) {
                sourceStates.remove(state);
            }
            return this;
        }

        public DefiningState onExit(final Action... action) {
            for (final T sourceState : sourceStates) {
                internals.addExitActions(sourceState, action);
            }
            return this;
        }

        public DefiningState onEntry(final Action... action) {
            for (final T sourceState : sourceStates) {
                log.debug("Create entry action for {} ({})", sourceState, action);
                internals.addEntryActions(sourceState, action);
            }
            return this;
        }

        public DefiningState isAnEndState() {
            for (final T state : sourceStates) {
                internals.addEndState(state);
            }
            return this;
        }

        public DefiningState isAStartState() {
            for (final T state : sourceStates) {
                internals.addStartState(state);
            }
            return this;
        }

        public DefiningState areEndStates() {
            return isAnEndState();
        }

        public DefiningState areStartStates() {
            return isAStartState();
        }

        @SafeVarargs
        public final DefiningTransition when(final Condition<E>... condition) {
            requireNonNull(condition);
            return new DefiningTransition(sourceStates, new Conditions<>(condition), internals);
        }

        // This method exists only to suppress warnings about varargs.
        public DefiningTransition when(final Condition<E> condition) {
            requireNonNull(condition);
            return new DefiningTransition(sourceStates, new Conditions<>(condition), internals);
        }

        @SafeVarargs
        public final DefiningTransition when(final E... events) {
            return new DefiningTransition(sourceStates, is(events), internals);
        }

    }

    public class DefiningTransition {
        private final Conditions<E> conditions;
        private final Actions actions = new Actions();
        private final Set<T> sourceStates;
        private P priority = defaultPriority;
        private final StateMachine<T, E, P>.Internals internals;

        public DefiningTransition(Set<T> sourceStates, Conditions<E> conditions, StateMachine<T, E, P>.Internals internals) {
            this.sourceStates = sourceStates;
            this.conditions = conditions;
            this.internals = internals;
        }

        public DefiningTransition action(Action action) {
            requireNonNull(action);
            this.actions.add(action);
            return this;
        }

        public DefiningState then(T destinationState) {
            requireNonNull(destinationState);
            return transition(destinationState, conditions, priority, actions);
        }

        public DefiningState transition(T destinationState, Conditions<E> storedConditions2, P priority, Actions actions) {
            this.actions.add(actions);
            for (var sourceState : sourceStates) {
                internals.addTransition(new Transition<>(sourceState, destinationState, storedConditions2, priority, this.actions));
            }
            return new DefiningState(sourceStates, internals);
        }

        public DefiningState transition(T destinationState, Condition<E> condition, P priority, Action... actions) {
            return transition(destinationState, new Conditions<>(condition), priority, new Actions(actions));
        }

        public DefiningTransition withPrio(final P priority) {
            requireNonNull(priority);
            this.priority = priority;
            return this;
        }
    }

    private StateMachine<T, E, P> machine;
    private final P defaultPriority;

    public DslStateMachineBuilder(P defaultPriority) {
        requireNonNull(defaultPriority);
        this.defaultPriority = defaultPriority;
    }

    @Override
    public StateMachine<T, E, P> build(StateMachine<T, E, P> newMachine) {
        Objects.requireNonNull(newMachine);
        machine = newMachine;
        executeBuildInstructions();
        return machine;
    }

    protected abstract void executeBuildInstructions();

    @Override
    public StateMachine<T, E, P> build() {
        return build(new StateMachine<>());
    }

    public DefiningState state(T state) {
        Objects.requireNonNull(state);
        return states(state);
    }

    @SafeVarargs
    public final DefiningState states(T... states) {
        return new DefiningState(new HashSet<>(Arrays.asList(states)), machine.new Internals());
    }

    public static <E> Condition<E> always() {
        return new AlwaysCondition<>();
    }

    public static <E> Condition<E> never() {
        return new NeverCondition<>();
    }

    public static <E> Condition<E> after(final long milliseconds) {
        return new AfterCondition<>(milliseconds);
    }

    @SafeVarargs
    public static <E> Conditions<E> is(final E... events) {
        Objects.requireNonNull(events);
        assert events.length != 0;

        if (events.length == 1) {
            var singleEvent = events[0];
            return new Conditions<>(new SingleEventMatchCondition<>(singleEvent));
        }

        return new Conditions<>(new MultiEventMatchCondition<>(events));
    }

    @SafeVarargs
    public final Condition<E> active(T... statesThatMustBeActive) {
        return new StatesActiveCondition<>(machine, statesThatMustBeActive);
    }

    @SafeVarargs
    public final Condition<E> inactive(T... statesThatMustBeInactive) {
        return new StatesInactiveCondition<>(machine, statesThatMustBeInactive);
    }

    public static Action log(final String logText) {
        Objects.requireNonNull(logText);
        return new LogAction(logText);
    }
}
